### DISCUSSION TITLE
sort_tables_and_constraints not working?

### DISCUSSION THREAD

**Discussion by FredericoCoelhoNunes** [sqlalchemy/sqlalchemy#6798] [Usage Questions] [answered]
Hi all,

I have a series of tables that I am trying to recreate in a different database. My idea is to iterate over these tables and then just do table.create(bind=new_engine) to create them
When I tried doing this, it threw an error because there are circular dependencies between foreign keys in a couple of the tables. Meaning, table A references table B, and vice-versa. This means that the tables need to be created first, and then "ALTERed" to create the foreign keys.

I have found a function - **sort_tables_and_constraints** - which I believe returns tables/constraints in an order such that they can be created one by one without conflicts.

Here's the description of the function:
```
This is a dependency-ordered sort which will emit tuples of (Table, [ForeignKeyConstraint, ...]) such that each Table follows its dependent Table objects. Remaining ForeignKeyConstraint objects that are separate due to dependency rules not satisfied by the sort are emitted afterwards as (None, [ForeignKeyConstraint ...]).
```

I am doing the following:
```
tables = [v for k, v in oracle_metadata.tables.items()]
for t, fks in sqlalchemy.schema.sort_tables_and_constraints(tables):
    if t is not None:
        if type(t) == sqlalchemy.sql.schema.Table:

            for c in t.columns:  # converting some types to be compatible with the new database
                if str(c.type) == 'NUMBER':
                    c.type = c.type.as_generic()  

            t.create(bind=mariadb_engine)
            
    if fks is not None:
        for fk_constraint in list(fks):
            fk_constraint.create(bind=mariadb_engine)
```

However, this still throws an error (this time it's a "malformed foreign key error" because some of the tables try to create foreign keys before the table they are referencing exists).

Any help would be much appreciated, I've been banging my head against the wall for an entire day!
Thanks

**Comment by zzzeek** [comment, marked as answer]
you can execute the `CreateTable` construct while omitting FK constraints:

```py
from sqlalchemy.sql.schema import CreateTable
with engine.begin() as conn:
    conn.execute(CreateTable(my_table, include_foreign_key_constraints=False))
```

**Reply by zzzeek** [reply]
although overall it looks like you're trying to do what I answered in https://github.com/sqlalchemy/alembic/discussions/876 .   I would stick to using `metadata.create_all`, just change the datatypes when you reflect them using the column_reflect event mentioned in that other answer.

**Reply by FredericoCoelhoNunes** [reply]
Hi,

thanks a lot for all your help! I could've sworn I had already tried that and ran into the foreign key errors, but apparently not - I am only running into the original error I had, which is: for some reason one of the tables has a duplicate constraint.

If you could validate my approach, I would be very grateful!

One of the source tables (Oracle) has the following constraint:
`CONSTRAINT "JHIST_DATE_INTERVAL" CHECK (end_date > start_date) ENABLE, `

It only shows up once in the table DDL.

When I reflect this table with SQLAlchemy, this constraint appears duplicated:
```
constraint_names = [c.name for c in problem_table.constraints]
print(constraint_names)
---> ['jhist_emp_id_st_date_pk', **'jhist_date_interval'**, 'jhist_dept_fk', 'jhist_job_fk', **'jhist_date_interval'**, 'jhist_emp_fk']
```

I am not sure if this is a bug, but as a fix, I was doing the following for all my tables to keep only one constraint per unique name:
```
for t in tables:
    constraint_names = set([c.name for c in t.constraints])

    if len(constraint_names) != len(t.constraints):
        print(t.constraints)
        # need to manually fix it, keeping one of the constraints in 
        new_constraints = []
        for name in constraint_names:
            for const in t.constraints:
                if const.name == name:
                    new_constraints.append(const)
                    break

        new_constraints = set(new_constraints)

        t.constraints = new_constraints
```

Any insights would be greatly appreciated @zzzeek , but wither way, you've already helped me A LOT.
Thanks!

**Reply by FredericoCoelhoNunes** [reply]
Also, one final question (I hope)!
If I use create_all, won't I run into problems inserting the data afterwards? Since both tables have foreign key constraint to each other, when I am trying to insert some data, it throws an error because the key doesn't exist in the other table.

Thanks!

**Reply by zzzeek** [reply]
when using relational databases, this is true that if you are INSERTing rows, they have to be INSERTed in the correct order.   if you have mutually dependent foreign keys then you will have to run in one of the values as NULL during INSERT, then on a second pass run an UPDATE to update it to the final value.    or if this is Oracle, maybe you can try using deferrable constraints, which wouldnt take effect until COMMIT.  Finally I think Oracle also allows a way of temporarily disabling constraints also, that is another route you can go if you are doing a one-time bulk insert.

**Reply by FredericoCoelhoNunes** [reply]
Thanks!
This is actually a migration from Oracle to MariaDB. I am now trying to do what you suggested and use the CreateTable function to create all tables omitting the constraints, then I will insert all data exactly as it is in the source database. and then I will apply the constraints. However I am getting the following error:

```
~/.virtualenvs/auchan/lib/python3.9/site-packages/sqlalchemy/sql/compiler.py in create_table_constraints(self, table, _include_foreign_key_constraints, **kw)
   4196         all_fkcs = table.foreign_key_constraints
   4197         if _include_foreign_key_constraints is not None:
-> 4198             omit_fkcs = all_fkcs.difference(_include_foreign_key_constraints)
   4199         else:
   4200             omit_fkcs = set()

TypeError: 'bool' object is not iterable
```

I think I will try to find a third-party pre-built solution for this database migration, as I can't seem to be able to figure it out! Cheers.
